src/server/models/Player.js

```diff
@@ -333,19 +341,18 @@ export default class Player extends Model {
 
   async discard(card, from = this.hand) {
     this.game.log(`${this.name} discards ${card.name}`);
-    const event = await this.handleTriggers('discard', { card, from }, [card]);
+    const event = await this.handleTriggers('discard', { cards: [card], from }, [card]);
     if (!event.handledByPlayer.get(this)) {
       this.moveCard(card, from, this.discardPile);
     }
   }
 
-  async discardAll(pile) {
-    const num = pile.length;
-    while (pile.length > 0) {
-      const card = pile.last();
-      const event = await this.handleTriggers('discard', { card, pile }, [card]);
-      if (!event.handledByPlayer.get(this)) {
-        this.moveCard(card, pile, this.discardPile);
+  async discardAll(cards, from = this.hand) {
```

Comment:
Yes, this is exactly the thing I've been thinking about making to handle cards like scheme and Herbalist. We'll have to discard hand and play area together in cleanup with this since they all technically get discarded at the same time.
I'll have to make a similar function for trashing, when trashing entire hand with count we don't want to get messed up if our hand changes half way through (fortress, hovel, rats)

Reply:
still need const num = cards.length;